R/oauth-init.R

```diff
@@ -82,28 +86,35 @@ init_oauth2.0 <- function(endpoint, app, scope = NULL, user_params = NULL,
     state <- nonce()
   }
 
-  authorize_url <- modify_url(endpoint$authorize, query = compact(list(
-    client_id = app$key,
-    scope = scope,
-    redirect_uri = redirect_uri,
-    response_type = "code",
-    state = state)))
-
-  if (use_oob) {
-    code <- oauth_exchanger(authorize_url)$code
+  # Some Oauth2 grant type not required an authorization request and code
+  # (see https://tools.ietf.org/html/rfc6749#section-4.4)
+  if(client_credentials) {
+    code <- NULL
   } else {
-    code <- oauth_listener(authorize_url, is_interactive)$code
+    authorize_url <- modify_url(endpoint$authorize, query = compact(list(
+      client_id = app$key,
+      scope = scope_arg,
+      redirect_uri = redirect_uri,
+      response_type = "code",
+      state = state)))
+
+    if (use_oob) {
+      code <- oauth_exchanger(authorize_url)$code
+    } else {
+      code <- oauth_listener(authorize_url, is_interactive)$code
+    }
   }
-
   # Use authorisation code to get (temporary) access token
 
   # Send credentials using HTTP Basic or as parameters in the request body
   # See https://tools.ietf.org/html/rfc6749#section-2.3 (Client Authentication)
+
   req_params <- list(
     client_id = app$key,
-    redirect_uri = redirect_uri,
-    grant_type = "authorization_code",
-    code = code)
+    redirect_uri = if (client_credentials) NULL else redirect_uri,
```

Comment:
Taking care of your previous review, I discovered the use of `httr:::compact`. it helps suppress `NULL` element in a list. I was wondering if we should add `compact(req_params)` after this as we are setting `redirect_uri` to `NULL` if `client_credentials` is `TRUE`